src/execution/execute.ts

```diff
@@ -826,32 +825,73 @@ function completeAbstractValue(
 
   return completeObjectValue(
     exeContext,
-    ensureValidRuntimeType(
-      runtimeType,
-      exeContext,
-      returnType,
-      fieldNodes,
-      info,
-      result,
-    ),
+    runtimeType,
     fieldNodes,
     info,
     path,
     result,
   );
 }
 
+function resolveType(
+  exeContext: ExecutionContext,
+  returnType: GraphQLAbstractType,
+  abstractType: GraphQLAbstractType,
+  fieldNodes: ReadonlyArray<FieldNode>,
+  info: GraphQLResolveInfo,
+  result: unknown,
+  encounteredTypeNames: Set<string> = new Set(),
+): GraphQLObjectType | Promise<GraphQLObjectType> {
+  const resolveTypeFn = abstractType.resolveType ?? exeContext.typeResolver;
+  const contextValue = exeContext.contextValue;
+  const possibleRuntimeType = resolveTypeFn(
+    result,
+    contextValue,
+    info,
+    abstractType,
+  );
+
+  if (isPromise(possibleRuntimeType)) {
+    return possibleRuntimeType.then((resolvedPossibleRuntimeType) =>
```

Comment:
Do you really need to add `resolveType` function to implement this feature or it's just refactoring mixed in?

The problem here is that adding one more `then` looks unnecessary.
Every `then` adds a trip to the microtasks queue even if the promise is already resolved and that significantly affects performance.
Because of that code, some time can look repetitive but it is done on purpose to minimize the number of `then` and improve performance.

Reply:
@IvanGoncharov, I totally agree with you, but that leaves me a bit puzzled.

This code actually is repetitive; I switch on whether the type resolver for the given abstract type (or the default type resolver) returned a promise, and only call `.then` when it is a promise. For exactly the reason you mention above, to not use then unnecessarily.

For that reason, I did need the `resolveType` function -- or, at least I thought I did.

Maybe you could be more specific about where you think the extra trip to the microtask queue is happening?

Reply:
feel free to directly modify this PR if you have something else in mind.

Reply:
summary of offline conversation:

> I am checking if we have a promise appropriately at each resolveType call, but because I refactored a recursive function that returned a PromiseOrValue, when there is no extra recursive steps, I ended up calling `then` twice: once for the only call to the possibly async type resolver, but also once on the entire recursive function.
>
> The recursion can be refactored (actually much closer to the original) so that if there is no need for recursion, the recursive part is not called.